src/ui/components/SecondaryToolbox/TerminalContextAdapter.tsx

```diff
@@ -20,16 +20,22 @@ export default function TerminalContextController({ children }: PropsWithChildre
   const idCounterRef = useRef(0);
 
   const addMessage = useCallback((partialTerminalExpression: NewTerminalExpression) => {
+    const terminalExpression: TerminalExpression = {
+      ...partialTerminalExpression,
+      id: idCounterRef.current++,
+      type: "TerminalExpression",
+    };
+
     // New expressions should be added in a transition because they suspend.
     startTransition(() => {
-      setMessages(prev => [
-        ...prev,
-        {
-          ...partialTerminalExpression,
-          id: idCounterRef.current++,
-          type: "TerminalExpression",
-        },
-      ]);
+      setMessages(prev => {
+        // warning: this callback may be called multiple times
+        // for a single addMessages() call
+        if (prev.some(expression => expression.id === terminalExpression.id)) {
+          return prev;
+        }
+        return [...prev, terminalExpression];
+      });
```

Comment:
Ah! It seems like I misunderstood what you were saying the other day. I thought you said the `startTransition` callback was called multiple times, which should not happen.

The `setState` callback definitely _might_ be called multiple times. Particularly if React has other updates and it wants to "rebase" this one on top of the latest state. That's expected behavior for React and one of the reasons `setState` callbacks should not have side effects (so I guess, technically my bad for incrementing the counter that way).

The "fix" for this should just be:

```js
const addMessage = useCallback(
  (partialTerminalExpression: NewTerminalExpression) => {
    // New expressions should be added in a transition because they suspend.
    startTransition(() => {
      const id = idCounterRef.current++;

      setMessages((prev) => [
        ...prev,
        {
          ...partialTerminalExpression,
          id,
          type: "TerminalExpression",
        },
      ]);
    });
  },
  []
);
```